### ISSUE TITLE
Feature request: sandbox.createStubInstance()

### ISSUE THREAD

**Issue by dpmott** [sinonjs/sinon#1598]
* Sinon version : 4.0.1
* Environment   : Windows 10
* Example URL   : 
* Other libraries you are using: Karma, Jasmine, Typescript

**What did you expect to happen?**
I was hoping to find an easy way to reset all of the stubs in an object returned by createStubInstance().
Given that the sandbox mirrors most methods that of stubs and spies, I was hoping to find a method like sandbox.createStubInstance() that would create an object whose stubbed methods were tied to the sandbox.

**What actually happens**
sinon.createStubInstance() creates independent stubs for each method, which appear to not be associated with any sandbox to which I have access.  There is no analogous sandbox method.

**How to reproduce**

I've made a utility helper function in my own code to achieve this:

```typescript
export const createStubInstance = (constructor: any, sandbox: any): any => {
  const stub = sinon.createStubInstance(constructor);
  for (const property in stub) {
    if (stub.hasOwnProperty(property)) {
      stub[property] = sandbox.stub();
    }
  }

  return stub;
};
```

Then, whenever I call `sandbox.reset()`, all of the stubs in the object created by the call to createStubInstance() will be reset (for behavior and history).

This lets me write cleaner code:

```js
  const sandbox = sinon.sandbox.create();
  const stubNgbModal = createStubInstance(NgbModal, sandbox);

  beforeEach(() => {
      stubNgbModal.open.returns({ result: Promise.resolve() });
  });

  afterEach(() => {
    sandbox.reset();
  });
```

whereas before, this code would either have assigned a new sandbox.stub() before every test:

```js
   stubNgbModal.open = sandbox.stub().returns({ result: Promise.resolve() }); // another allocated stub
```

or the afterEach would need to clean up the stub by-name:

```js
   stubNgbModal.open.reset(); // a line like this for every customized stub
```

**What now?**
I'm rather a novice with using Sinon; if there's a better way to achieve what I'm trying to do, I'd love to hear about it.

If there's interest in this feature request, I'll spend some time to create a pull request.

Thanks for reading!

**Comment by mroderick** [comment]
I think that having `createStubInstance` on `sandbox` would be a natural improvement. A pull request that implements this would be most welcome!

In your helper you're using [`for..in`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in) and [`hasOwnProperty`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty). Sinon uses ES5.1 which includes [`Object.keys`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys) and [`Array.prototype.forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach), which would be preferred over `for...in` and `hasOwnProperty`.

Your PR should include an update to the documentation in `docs/release-source/release/sandbox.md` also.

Looking forward to your PR 🤸‍♂️

**Comment by mroderick** [comment]
This was implemented with #1602 and published to npm as `sinon@4.1.0`
